### ISSUE TITLE
File transfer in history

### ISSUE THREAD

**Issue by sphaerophoria** [qTox/qTox#5350]
As part of \#5348 I've started looking into how to implement file transfer history. I figure this won't be wasted work as even if \#5348 doesn't go through this is still a useful change.

I'm opening this ticket instead of just a PR as I'd like to discuss database options. I currently have very little context on the project so input from people with more background knowledge would be greatly appreciated.

This change should also help pave the way for resuming file transfers after qtox restart.

In order to assist discussion of DB layout I've summarized a bit of what I've learned about file transfers below.

### How our current chat form handles signals from Core

This part isn't super important for database layout but will end up being helpful to reference later when hooking up signals to History()

Signal|Trigger|Action
------|-------|------
fileSendStarted|User initiates a file transfer with friend|Create a new file transfer message and insert it into the chat log. Update friend activity (probably to double check they're online?)
fileSendFailed|toxcore returned -1 from tox_file_send. Note that this isn't a reliable return value. tox.h indicates that on failure it returns an "unspecified value", however it seems that this reliably happens if we try to send a file and the user is offline. As an aside we should use the error param here..| Status message inserted saying "Failed to send file" 
fileReceiveRequested|Friend initiates a file transfer|Inserts a new file transfer message into the chatlog, if autoaccept is enabled accept the file transfer
fileTransferAccepted|We accept a file transfer from a friend (not the other way around) **or** we unpause a file transfer|Set the background of the file to light grey and re-init the transfer button.
fileTransferCanceled|Any situation where the file transfer stopped mid way. This seems to be emitted if a user cancels the transaction **or** if there was internal error|Color the background red and hide widgets as this is now a **finished** file transfer
fileTransferFinished|File transfer completed sucessfully|Setup file actions (open file, open dir) and color widget green as this is a **finished** file transfer
fileUploadFinished|fileTransferFinished but only for uploads|Adds file transfer to the file transfer view for uploads
fileDownloadFinished|fileTransferFinished but only for downloads|Adds file transfer to the file transfer view for downloads
fileTransferPaused|We paused a file transfer (either an outgoing or incoming one)|Set color of widget to light grey and reinitialize our buttons
fileTransferInfo|Callback from toxcore when, emitted when we're still sending a file to update progress|Update our progress bar
fileTransferRemotePausedUnpaused|Remote sent a pause/unpause signal|We do the same thing as when we pause/resume (accept) transfers. Set color to grey and reinit buttons
fileTransferBrokenUnbroken|This seems to happen when we notice that a friend has gone offline/online half way through a transfer|Currently we just pretend the file transfer was canceled if it was broken and do nothing when they come back, however it looks like this was implemented in a way to let us resume file transfers when they come back online at a later date
fileNameChanged|this comes in through the tox file receive callback. If the file request has an invalid filename we sanitize it before emitting this signal|Display a system warning that we changed the filename compared to what the friend actually sent

### Current toxfile data

```
    uint8_t fileKind;
    uint32_t fileNum;
    uint32_t friendId;
    QByteArray fileName;
    QString filePath;
    std::shared_ptr<QFile> file;
    quint64 bytesSent;
    quint64 filesize;
    FileStatus status;
    FileDirection direction;
    QByteArray avatarData;
    QByteArray resumeFileId;
```

### Resuming transfers accross restarts

This part I'm pretty shakey on since I haven't tried it, but it looks to me like the important data to save here is how much data has been transfered if we're the receiver, and the unique file_id retrieved with tox_file_get_file_id(). From inspection of tox.h I believe this is the flow to resue a file transfer after a toxcore restart

* Sender initiates transfer with tox_file_send()
* Recipient calls their file_recv callback
* Recipient calls tox_file_seek()
* Recipient calls tox_file_control() with TOX_FILE_RESUME

### Data we need to track from ToxFile

```
    uint8_t fileKind; --> No, this is either avatar or file and we only care about history for files
    uint32_t fileNum; --> No, even in the resume file transfer case I believe we'd use the file_id field instead
    uint32_t friendId; --> Yes, which friend's chatroom will we show it in
    QByteArray fileName; --> Yes, we'll need this to re-construct the file transfer widget
    QString filePath; --> Yes, we'll need this to find the file to display in the transfer widget, also to re-initiate file transfers
    std::shared_ptr<QFile> file; --> No don't need to store this, we'll re-open the file when we load
    quint64 bytesSent; --> No, assuming my understanding of resuming file trasnfer after restart is correct, only the reciever needs to keep track of transfer state after restart, and we can just read the file data on disk for that
    quint64 filesize; --> Yes(?). I believe that after a restart to resume file transfers we will need to check if our on disk size matches our reported size
    FileStatus status; --> I don't think we care about this, after restart this should always be STOPPED 
    FileDirection direction; --> Yes, we'll need this to determine who sent it in the chatlog
    QByteArray avatarData; --> No, we don't care about the avatar type
    QByteArray resumeFileId; --> Yes, this will be needed to re-init file transfers
```

### Database layout proposal

* Add a file_transfer_id column to the history table. I think we can re-use the persistant transfer id.
* add a file_transfer table with the information marked as needed above plus a flag indicating whether or not we've completed the transfer
    * table colums: id (autogenerated by sqlite), file_id, friend_id, file_name, file_path, file_size, direction, transfer_finished

I believe the above should be enough to show in our chat log *and* resume transfers. On load of a chatroom we can for the time being display unfinished transfers as failed. In the future we can use the file_id + file_size to resume the transfer

If our transfer is finished friend_id, file_name, file_path, and direction should be enough to reconstruct the file widget and initialize image previews etc.

Can you guys think of anything I've missed?

## Potentially discovered bugs during investigation

Also while browsing I noticed the following oddities, could a maintainer let me know if I should be filing tickets for them?

* We do not use the error field of tox_file_end but instead use a return value that tox.h indicates we shouldn't use
* ~~We emit fileTransferPaused / fileTransferAccepted before actually telling toxcore to do it. This means that if we were really slow in any of our handlers we would keep putting data on the wire when the user didn't want us to~~ On second thought I think this actually reasonable. If the tox functions take a long time to respond it probably makes sense to update our UI before actually executing the call.
* ~~I think we might call QFile::setFileName after already opening the file. in CoreFile::acceptFileReceiveRequest we open a file with the name given to us by toxcore. On the first chunk received in CoreFile::onFileReceiveCallback we sanitize the file path with CoreFile::getCleanFileName and then set the path again. According to the Qt Docs this is invalid API usage "Do not call this function if the file has already been opened" http://doc.qt.io/qt-5/qfile.html#setFileName~~
I'm sorry I mis-diagnosed order here. onFileReceiveCallback is called on file transfer request.

**Comment by sphaerophoria** [comment]
After playing a little more catch up on databases I've modified my proposal to not use the file transfer id as the primary key for the database.

**Comment by sudden6** [comment]
I have not yet gone through the whole table summarizing the status quo, but here are my first thoughts anyway.

First of all, the file transfer stuff in qTox is quite brittle and not fully implemented so please don't take too much of the code as a good example ;)

> uint32_t friendId; --> Yes, which friend's chatroom will we show it in

`friendId` is not persistent, but per session i.e. after a restart it's lost. So it's not a good thing to reference the chatroom. We try to move everything to the public key of the friend, which is persistent and guaranteed to be unique.

As a side thought, `resumeFileId` seems to be perfectly suited to be a hash of the sent file, but I found no place in toxcore that this is guaranteed, maybe we can propose that to TokTok/c-toxcore?

> add a file_transfer table with the information marked as needed above plus a flag indicating whether or not we've completed the transfer

    table colums: id (autogenerated by sqlite), file_id, friend_id, file_name, file_path, file_size, direction, transfer_finished

- As said above, `friend_id` should be replaced by `friend_pk`. 
- what do you think about making `transfer_finished` a field that can also capture the `Paused` and `Aborted` state of a file transfer?
- Do we need a field that stores the number of bytes received in order to resume file transfers?

> Potentially discovered bugs during investigation

I think those are all real bugs :(

**Comment by sphaerophoria** [comment]
Thanks for the input.

I was thinking we could store the friend_id as a reference to the peer database table, which is resolvable back to a friend public key after restart. This is what we do already for other history messages (although friend id is an overloaded term so it was completely unclear above :P )

I like the idea of transfer finished becoming transfer state with more info.

I *think* that the reciever can open the file and check the size on next boot instead of storing it in the db.

Ill split off issues for the discovered bugs

**Comment by sudden6** [comment]
> I was thinking we could store the friend_id as a reference to the peer database table, which is resolvable back to a friend public key after restart. 

Ok, didn't know there was already a table with all the public keys.

> I think that the reciever can open the file and check the size on next boot instead of storing it in the db.

Hmm, this would also make sense.

**Comment by sphaerophoria** [comment]
Sorry about the mis-diagnosed issues. I misread what onFileRecvCallback was for, this actually comes in before we receive any data. I thought this was when we received a chunk of a file. Submitted PR for the tox file api misue, and opted not to change behavior of emitting signals before sending down to toxcore. If toxcore takes a while I think it actually makes sense for our UI to update first.

**Comment by sphaerophoria** [comment]
> > I think that the reciever can open the file and check the size on next boot instead of storing it in the db.
> 
> Hmm, this would also make sense.

This also has the benefit of not hammering the database on each chunk receive.

**Comment by anthonybilinski** [comment]
Closed by #5354.
